File: yuv/src/main/java/com/theeasiestway/yuv/FramesFactory.kt
package com.theeasiestway.yuv

import com.theeasiestway.yuv.entities.ArgbFrame
import com.theeasiestway.yuv.entities.YuvFrame
import java.nio.ByteBuffer

//
// Created by Loboda Alexey on 25.06.2020.
//

object FramesFactory {

    fun instanceYuv(dstWidth: Int, dstHeight: Int): YuvFrame {
        val yuvFrame = YuvFrame()
        val ySize = dstWidth * dstHeight
        val uvSize = dstWidth * dstHeight / 4
        val y = ByteBuffer.allocateDirect(ySize)
        val u = ByteBuffer.allocateDirect(uvSize)
        val v = ByteBuffer.allocateDirect(uvSize)
        val extra = if (dstWidth % 2 == 0) 0 else 1
        yuvFrame.fill(y, u, v, dstWidth, dstWidth / 2 + extra, dstWidth / 2 + extra, dstWidth, dstHeight)
        return yuvFrame
    }

    fun instanceYuv(width: Int, height: Int, rotationMode: Int): YuvFrame {
        val newFrame = YuvFrame()
        val outWidth = if (rotationMode == 90 || rotationMode == 270) height else width
        val outHeight = if (rotationMode == 90 || rotationMode == 270) width else height
        val ySize = outWidth * outHeight
        val uvSize = outWidth * outHeight / 4
        val y = ByteBuffer.allocateDirect(ySize)
        val u = ByteBuffer.allocateDirect(uvSize)
        val v = ByteBuffer.allocateDirect(uvSize)
        val extra = if (outWidth % 2 == 0) 0 else 1
        newFrame.fill(y, u, v, outWidth, outWidth / 2 + extra, outWidth / 2 + extra, outWidth, outHeight)
        return newFrame
    }

    fun instanceArgb(width: Int, height: Int): ArgbFrame {
        val newFrame = ArgbFrame()
        val data = ByteBuffer.allocateDirect((width * height) * 32)
        val extra = if (width % 2 == 0) 0 else 1
        val dataStride = width * 4 + extra
        newFrame.fill(data, dataStride, width, height)
        return newFrame
    }
}

File: yuv/src/main/cpp/factories/FramesFactory.cpp
//
// Created by Loboda Alexey on 29.07.2020.
//

#include "FramesFactory.h"
#include "../entities/YuvFrame.h"
#include "../entities/RgbFrame.h"
#include "../LibyuvWrapper.h"

YuvFrame* FramesFactory::instanceYuv(int width, int height) {
    int ySize = width * height;
    int uvSize = ySize / 4;
    int extra = width % 2 == 0 ? 0 : 1;
    return new YuvFrame(width, height, width, width / 2 + extra, width / 2 + extra, ySize, uvSize, uvSize);
}

YuvFrame* FramesFactory::instanceYuv(int width, int height, int rotation) {
    int outWidth = (rotation == 90 || rotation == 270) ? height : width;
    int outHeight = (rotation == 90 || rotation == 270) ? width : height;
    int ySize = outWidth * outHeight;
    int uvSize = outWidth * outHeight / 4;
    int extra = outWidth % 2 == 0 ? 0 : 1;
    return new YuvFrame(outWidth, outHeight, outWidth, outWidth / 2 + extra, outWidth / 2 + extra, ySize, uvSize, uvSize);
}

RgbFrame* FramesFactory::instanceArgb(int width, int height) {
    int dataSize = (width * height) * 4; // 4 bytes: (8 + 8 + 8 + 8)bits / 8 bits
    int extra = width % 2 == 0 ? 0 : 1;
    int stride = width * 4 + extra;
    return new RgbFrame(width, height, dataSize, stride, LibyuvWrapper::ARGB);
}

RgbFrame* FramesFactory::instanceRgb565(int width, int height) {
    int dataSize = (width * height) * 2; // 2 bytes: (5 + 6 + 5)bits / 8 bits
    int extra = width % 2 == 0 ? 0 : 1;
    int stride = width * 2 + extra;
    return new RgbFrame(width, height, dataSize, stride, LibyuvWrapper::RGB565);
}